src/vision/shiny_future/alan_switches_runtimes.md

```diff
@@ -0,0 +1,148 @@
+# ✨ Shiny future stories: template
+
+[How To Vision: Shiny Future]: ../how_to_vision/shiny_future.md
+[the raw source from this template]: https://raw.githubusercontent.com/rust-lang/wg-async-foundations/master/src/vision/shiny_future/template.md
+[`shiny_future`]: https://github.com/rust-lang/wg-async-foundations/tree/master/src/vision/shiny_future
+[`SUMMARY.md`]: https://github.com/rust-lang/wg-async-foundations/blob/master/src/SUMMARY.md
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "shiny future" story submitted as part of the brainstorming period. It is derived from what actual Rust users wish async Rust should be, and is meant to deal with some of the challenges that Async Rust programmers face today.
+
+If you would like to expand on this story, or adjust the answers to the FAQ, feel free to open a PR making edits (but keep in mind that, as peoples needs and desires for async Rust may differ greatly, shiny future stories [cannot be wrong]. At worst they are only useful for a small set of people or their problems might be better solved with alternative solutions). Alternatively, you may wish to [add your own shiny vision story][htvsq]!
+
+## The story
+
+Since his [early adventures](./alans_trust_in_the_compiler_is_rewarded.md) with Async I/O went so well, Alan has been looking for a way to learn more. He finds a job working in Rust. One of the projects he works on is [DistriData]. Looking at their code, he sees an annotation he has never seen before:
+
+```rust
+#[async_runtime(humboldt)]
```

Comment:
Some async runtime runs on single thread and some runs on multiple threads. How do we different among it if we do it this way? Also, will the internal code changes along they way? If we force everything to be Arc-wrapped wouldn't we impose an extra cost on those using it on a single thread?

Reply:
I don't quite follow. I am imagining something like this:

* If one writes `async fn main`, then the compiler generates a synchronous `main` function that invokes a method similar to `block_on` from some crate. You can specify the crate via the `async_runtime` annotation (or perhaps some other annotation; e.g., perhaps just the `#[humboldt::main]` convention that seems to be in wide use today).
* If one prefers, you can just write `fn main` yourself, which would allow you to instantiate runtimes with whatever options you like.

I can spell out some of this in a FAQ -- but I want to be sure first, does that address your point? It seems to me like it does. The runtime could indeed be single threaded or not, and have all kinds of other options besides.

Reply:
I'm going to change to `humboldt::main`

Reply:
Pushed some changes, @pickfire.

Reply:
No, that's not what I am talking about. I mean for example, `tokio::main` two different flavors `flavor = "multi-thread"` and `flavor = "current_thread"`, and also options for those flavors. How do we intend to support those with `#[async_runtime(...)]`? https://docs.rs/tokio/1.4.0/tokio/attr.main.html

Also, it is also common to see some runtime only supporting multi-thread such as async-std but others only supporting single-thread such as actix, and some such as tokio supports both. Is it possible to know how many threads it supports as in having an explicit information for this? And how can we pass in the additional information for tokio such as `worker_threads`?

But yeah, changing that to `#[humboldt::main]` seemed to fix the issue. Just that now the user needs to read the docs to know whether it is running with multi-thread or single-thread (like how it is done currently).

The other parts I mentioned can just ignore, it's probably not for this issue.

Reply:
Yeah, that's partly why I changed to `humboldt::main`, it gives more control to the crate to define exactly what options. I definitely agree that the user will have to read the executor to understand what it means to use it. I think that's fine. A more interesting question is how that affects the "interoperability" traits, which have to be targeted for maximum flexibility, but the point of the "shiny future" is not to have all the answers.